controllers/user.js

```diff
@@ -1,15 +1,132 @@
 // --- Routes
 
-router.get('/', (req, res) => res.send('Hello World!'));
-// post
-// put
-// delete
+// Endpoint:  /api/users
+//Creating a user
+router.post("/",async (req,res)=>{
+
+    const user = new User(req.body);
+    try{
+        const token = await user.newAuthToken()    //Generate auth token
+        res.status(201).send({user, token})     
+        //We also send the token along with the user so to identify which token is the user currently logged in with
+    }catch(e){
+        res.status(400).send(e)
+    }
+})
+
+//   Endpoint:  /api/users/login
+//Logging in users
+router.post("/login",async (req,res)=>{
+    try{
+        const user = await User.findOne({email:req.body.email})
+        if(!user){
+            throw new Error("Unable to login")
+        }
+        const isMatch = await bcrypt.compare(req.body.password,user.password);
+        if(!isMatch){
+            throw new Error("Unable to login")
+        }
+        //If email and password are correct,it generates a new token and appends it to the tokens array
+        const token = await user.newAuthToken()
+        res.send({
+            user:user,
+            token:token    
+        })
+    }catch(e){
+        console.log(e)
+        res.status(401).send(e)
+    }
+})
+
+
+//Endpoint: /api/users/logout
+//Route to logout user from his current session
+router.post('/logout', auth, async (req, res) => {
+    try {
+        req.user.tokens = req.user.tokens.filter((token) =>{
+         return token.token !== req.token 
+        })
+        await req.user.save()
+        res.send("Logged out from current session")
+    } catch (error) {
+        res.status(500).send()
+    }
+})
+
+
+
+//Endpoint:  /api/users/me
+//Gives details about the currently logged in user
+router.get("/me",auth,async (req,res)=>{
+    res.send(req.user) 
+})
+
+
+//Endpoint: /api/users/me
+//Route to update current user
+router.patch('/me', auth ,async (req,res) => {
+    console.log("Heree")
+    const updates  = Object.keys(req.body)      //Returns all the keya as an array
+    const allowedUpdates = ["name", 'username',"email", "password", "graduationDate","bio","currentSchool","accepts","rejects","pinnedQuestions"]
+    const isValidOperation = updates.every((update) => allowedUpdates.includes(update))
+
+    if(!isValidOperation){
+        res.status(400).send({error:'Invalid request'})
+    }
+
+    try {        
+        updates.forEach((update) => req.user[update] = req.body[update]) 
+        await req.user.save()
+        res.send(req.user);
+    } catch (error) {
+        res.status(400).send()
+    }
+
+})
+
+
+//Endpoint: /api/users/me
+//Route to delete current user
+router.delete('/me', auth, async (req,res) => {
+    try {
+        await req.user.remove()
+        res.send("User deleted")
+    } catch (error) {
+        res.status(500).send()
+    }
+})
+
+
+//Endpoint: /api/users/:id
+//Get user by id
+router.get('/:id', async (req,res) => {
```

Comment:
Just make sure you apply a check to find whether the user with the given credentials exists in the db or not before GETing the new user.